### PR TITLE
DAOS-623 client: allow variable declaration in loop

### PR DESCRIPTION
**Pull request by jxiong**
This PR tries to enable variable declaration in loop from C99 standard. This will not make coding easier, but will also make the code less error-prone by limiting the scope of looping variables.

The Linux kernel adopts C99 for this exact same reason.

NB: I only updated the variable in the for loops. It seems like this file hasn't been changed since new clang-format is used. Maybe I picked a wrong file to start with ;-)

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by wangshilong — APPROVED** [review]
It might be better to have a ticket number, I thought there is a MISC ticket for such PRs. @jolivier23

**Comment by jolivier23** [comment]
> It might be better to have a ticket number, I thought there is a MISC ticket for such PRs. @jolivier23

there is DAOS-623

**Review by jolivier23 — DISMISSED** [review]
it would appear you don't need to change anything in clang-format. It must not care with current settings.

**Comment by jxiong** [comment]
> It might be better to have a ticket number, I thought there is a MISC ticket for such PRs. @jolivier23

@wangshilong This PR is just for reference purpose so that I can collect community's feedback. I will certainly add a daos ticket once we decide to move it forward. Thanks for calling it out.

**Review by ashleypittman — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Comment by jxiong** [comment]
@jolivier23 I don;t think this CI test (continuous-integration/jenkins/pr-head) would pass, I have tried 3 times but no luck. What would be your suggestion for the next step?

**Comment by ashleypittman** [comment]
> @jolivier23 I don;t think this CI test (continuous-integration/jenkins/pr-head) would pass, I have tried 3 times but no luck. What would be your suggestion for the next step?

There was a network issue when the build was starting, I've retriggerd it.

**Comment by daltonbohning** [comment]
Ideally, we should get a review from @daos-stack/client-api-owners before merging.

File: src/client/api/object.c
Comment (mchaarawi) [inline src/client/api/object.c:314]: this is not our coding style

**Review by mchaarawi — CHANGES_REQUESTED** [review]
i do not get the purpose of doing these changes on just this file. this screws up git blame.
we probably have many other files that do not have clang format and i will request we don't do such changes for now unless we make a discussion to address this everywhere.

**Comment by jxiong** [comment]
> i do not get the purpose of doing these changes on just this file. this screws up git blame. we probably have many other files that do not have clang format and i will request we don't do such changes for now unless we make a discussion to address this everywhere.

I agree with you that we should not make PRs for this change only. This is just to make an example of such change. The PR description gives a clearer explanation on this. The goal is sort of to announce that this is a preferred way of defining loop variables.

**Comment by mchaarawi** [comment]
> > i do not get the purpose of doing these changes on just this file. this screws up git blame. we probably have many other files that do not have clang format and i will request we don't do such changes for now unless we make a discussion to address this everywhere.
> 
> I agree with you that we should not make PRs for this change only. This is just to make an example of such change. The PR description gives a clearer explanation on this. The goal is sort of to announce that this is a preferred way of defining loop variables.

i think step 1 would be to revert all the clang format changes in the file. you should not format code the you did not touch. I am not ready to loose git blame information on all files :-) 
We use git blame to assign issues in some situations.

for the loop variable declaration, this is not something we discussed or adopted, so we cannot force it on everyone without a discussion in gatekeeping at least.

**Comment by jxiong** [comment]
> > > i do not get the purpose of doing these changes on just this file. this screws up git blame. we probably have many other files that do not have clang format and i will request we don't do such changes for now unless we make a discussion to address this everywhere.
> > 
> > 
> > I agree with you that we should not make PRs for this change only. This is just to make an example of such change. The PR description gives a clearer explanation on this. The goal is sort of to announce that this is a preferred way of defining loop variables.
> 
> i think step 1 would be to revert all the clang format changes in the file. you should not format code that you did not touch. I am not ready to loose git blame information on all files :-) We use git blame to assign issues in some situations.
> 
> for the loop variable declaration, this is not something we discussed or adopted, so we cannot force it on everyone without a discussion in gatekeeping at least.

That's absolutely reasonable. I don't have any background on this topic but it seems like we have a clang-format in place but it's not actually in use. (I have dev config that clang-format will be invoked automatically on file save).

I will revert the clang-format change and resubmit.

File: src/client/api/object.c
Comment (ashleypittman) [reply src/client/api/object.c:331]: We don't have much code that looks like this but keeping the scope of variables as small as possible is something that is explicitly stated in our style guide and it's something that's generally considered best-practice across the industry.  I know I've fixed at least one bug in our codebase where a loop variable was misused later in the function.

Changing existing code is laborious and time consuming, plus there are the issues it causes with glt blame and patches across branches etc but I that doesn't change the fact this style would be safer, more concise and preferable overall.

Comment (mchaarawi) [reply src/client/api/object.c:331]: i do not disagree. but we are not going to go on and change every part of the code to do this.
we are just going to cause more disruptions to ongoing prs, feature branches that is not going to give that much gain.

so i will approve the change for this file, because i know there is not much work going on in the object api area. @jxiong but please refrain from pushing such a change more broadly before discussing this more in developer channels and/or gatekeeping channels.

**Review by mchaarawi — APPROVED** [review]

File: src/client/api/object.c
Comment (ashleypittman) [reply src/client/api/object.c:331]: @mchaarawi are you in the community slack channel?  My understanding is this PR was created to support a discussion there.

**Review by ashleypittman — APPROVED** [review]

File: src/client/api/object.c
Comment (mchaarawi) [reply src/client/api/object.c:331]: I am. this is not a community question though. this is developer specific topic.

**Comment by jxiong** [comment]
@jolivier23 @ashleypittman can anyone of you please restart the CI for me? Thanks

**Comment by ashleypittman** [comment]
> @jolivier23 @ashleypittman can anyone of you please restart the CI for me? Thanks

You will need to merge with master to pickup the code fix, master was failing for a few days at the start of last week.  Master was fixed on Tuesday and Existing approvals should be maintained.

**Comment by jxiong** [comment]
NLT is broken. @daos-stack/daos-gatekeeper can we just land it as-is?

**Comment by mchaarawi** [comment]
Please refrain from using DAOS-623 in the future. thanks :-)

**Pull request merged**
